### PR TITLE
Improve binding element type inference

### LINKED ISSUE #49989: Using variable defined by array destructuring assignment in default value of variables defined afterwards.
# Bug Report

### 🔎 Search Terms

Array destructuring assignment default value 7022

### 🕗 Version & Regression Information

Exists in 3.3.3 - 4.8.0-beta
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about this behaviour

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=8&ts=4.8.0-beta#code/MYewdgzgLgBA2gQwDQwEYwLwwTA1DARgF1N4AmIgbgFgAoUSEAGwFMA6JkAcwAoADBAC4YAEgDeCAL4pUw8akl8AlDVpA)

### 💻 Code

```ts
const [a, b = a + 1] = [2];
console.log(`a: ${a}, b: ${b}`);
```

### 🙁 Actual behavior

Types of `a` and `b` inferred as `any`, or if `noImplicitAny` enabled, shows error: `'a' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)`

### 🙂 Expected behavior

Types of both `a` and `b` should be inferred as `number`.

### PR DESCRIPTION
**Pull request by babakks**
Fixes #49989

## Problem

There were two problems:

1. Type inference for variables declared by binding elements was not accurate. For example, for the declaration below the type of both variables were inferred to be `any` (instead of `number`).

   ```ts
   const [a, b = a] = [1];
   ```

1. With `noImplicitAny` option enabled, a false circular relationship error would be emitted: 
  
   ```text
   'a' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)
   ```

## Cause

The problems were hidden in the `getTypeFromBindingElement` function whose responsibility is to infer the type of variables defined as binding elements *without* looking onto the declaration initializer (the array literal, `[1]`, in the example above).

When a binding element itself had a default value (as in `b = a`) and that default value involved any sibling variable (i.e., variables defined in the same binding pattern, here `a`) then the inquiry for the type of the default value would begin and eventually result in a circular relationship error down the path. This error made the type checker to infer the types as `any`.

## Solution

Since the `getTypeFromBindingElement` function just needs to narrowly look inside the binding elements/patterns, we can skip further type checkings (that is falling back to `any`) when any other sibling variables are somehow involved in the default value expressions.

Although this seems to be an ad-hoc/one-off kind of solution, type checkings for binding elements are deeply integrated with other parts of the type checker, and solving the problem in a more generic way, I believe, requires greater knowledge of the code base (compared to mine, of course).

## Effect on the existing behavior

Three existing test cases were affected by this PR, fortunately in a good way. They're now correctly inferring the type `number`. Also, the circular relationship error was erased from one similar test case.

## New test case

A new test case, `destructuringArrayBindingPatternAndAssignment5.ts`, was added that contains cases where default values somehow reference other sibling variables.

### CODE REVIEW

File: tests/baselines/reference/useBeforeDeclaration_destructuring.types [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
 a;
->a : any
+>a : string
 
 let {a, b = a} = {a: '', b: 1};
->a : any
->b : any
->a : any
+>a : string
+>b : string | number
```
Comment (Andarist) [inline tests/baselines/reference/useBeforeDeclaration_destructuring.types:7]: I feel like this probably could be improved as it's impossible for `b` to be of type `string` here. Perhaps this doesn't have to be improved in this PR though and a followup issue could be created to track this.

Comment (babakks) [reply tests/baselines/reference/useBeforeDeclaration_destructuring.types:7]: Totally agree.

**Comment by andrewbranch** [comment]
@typescript-bot perf test this
@typescript-bot pack this

**Comment by andrewbranch** [comment]
I haven’t checked yet, but I believe this might break this case:

```ts
const [a, b = () => a] = [0]
```

I doubt that eagerly checking for circularities by syntax/symbol is going to be a workable approach because of cases like this. I think this example works today only because checking function bodies is a deferred operation, which may provide a hint for an alternative approach, but I’m really not sure. You can also imagine cases where `a` is referenced in `b`’s initializer but not in any way that contributes to the type, e.g.

```ts
const [a, b = (a, 0)] = [0]; // errors today
const [c, d => { console.log(c); return 0; }] = [0]; // ok
```

Are you sure that this code path doesn’t just need to pass `reportErrors` as `false` (or propagate that down into further checks, if it already is `false`)? If we’re just trying to get a contextual type for the variable declaration initializer, it doesn’t seem like errors should be reported. What is the call stack when the implicit any error is reported?

**Comment by andrewbranch** [comment]
It looks like the examples I mentioned still work in your PR—but that really still makes me think it should just be possible to squash the error without doing the eager circularity check.

**Comment by babakks** [comment]
@andrewbranch Thanks for the hints. Actually, this wasn't my first approach. I was looking for some way to suppress type-checking errors by propagating `reportErrors` just as you said, but, unfortunately, it'd go deep into a good number of methods. To be more specific, to propagate `reportErrors` argument, we have to add it as a parameter to the `checkExpression` and also other type-check methods called from that on (e.g., `checkIdentifier` in the example below). Obviously, changing the signature for all those methods wasn't a good solution. Neither defining a new state variable accessible to all those methods would incur other statefulness problems/corner cases. So, I tried to eliminate the unnecessary type-check with a rather repressive, ad-hoc approach which you see in this PR.

Below is the stack trace from the `getTypeFromBindingElement` down to the point where we wrongly push the symbol onto the type-checking stack (which shouldn't happen since we don't want to report errors yet). This session has been run on an existing test involving binding elements (i.e.,  [`destructuringArrayBindingPatternAndAssignment3.ts`][test-file]) on the `main` branch (at 9767f517):

[test-file]: https://github.com/microsoft/TypeScript/blob/9767f51758c2899ad09c3466fd6ebbde79f53f63/tests/cases/conformance/es6/destructuring/destructuringArrayBindingPatternAndAssignment3.ts

```ts
const [a, b = a] = [1]; // ok
...
```

As in the stack trace, there are 3 important points (marked with asterisks):

- The first one (*) is when the `getTypeFromBindingElement` method is called for the first time (I mean called on the second binding element, `b = a`) and `reportErrors` is expectedly set to `false`.
- The second one (**) where  `getTypeFromBindingElement` is called again (on the `b = a` element) with `reportErrors = false`.
- The last one (***) on the top line of the stack trace is where we pushed the symbol `a` on the type-checking stack, which shouldn't happen since we don't want errors reported.

With this PR, just the first call happens, and then the type checking returns with `any`.

## Stack trace

```text
*** at pushTypeResolution (/workspaces/TypeScript/built/local/run.js:56208:45)
    at getTypeOfVariableOrParameterOrPropertyWorker (/workspaces/TypeScript/built/local/run.js:57181:18)
    at getTypeOfVariableOrParameterOrProperty (/workspaces/TypeScript/built/local/run.js:57121:28)
    at getTypeOfSymbol (/workspaces/TypeScript/built/local/run.js:57493:24)
    at getNarrowedTypeOfSymbol (/workspaces/TypeScript/built/local/run.js:72421:20)
    at checkIdentifier (/workspaces/TypeScript/built/local/run.js:72497:24)
    at checkExpressionWorker (/workspaces/TypeScript/built/local/run.js:80745:28)
    at checkExpression (/workspaces/TypeScript/built/local/run.js:80694:38)
    at checkExpressionWithContextualType (/workspaces/TypeScript/built/local/run.js:80324:28)
    at checkDeclarationInitializer (/workspaces/TypeScript/built/local/run.js:80374:21)
**  at getTypeFromBindingElement (/workspaces/TypeScript/built/local/run.js:56990:81)
    at /workspaces/TypeScript/built/local/run.js:57040:109
    at [Object.map](http://object.map/) (/workspaces/TypeScript/built/local/run.js:455:29)
    at getTypeFromArrayBindingPattern (/workspaces/TypeScript/built/local/run.js:57040:35)
    at getTypeFromBindingPattern (/workspaces/TypeScript/built/local/run.js:57063:19)
    at getContextualTypeForInitializerExpression (/workspaces/TypeScript/built/local/run.js:73283:28)
    at getContextualType (/workspaces/TypeScript/built/local/run.js:73808:28)
    at getApparentTypeOfContextualType (/workspaces/TypeScript/built/local/run.js:73727:17)
    at checkArrayLiteral (/workspaces/TypeScript/built/local/run.js:74170:34)
    at checkExpressionWorker (/workspaces/TypeScript/built/local/run.js:80775:28)
    at checkExpression (/workspaces/TypeScript/built/local/run.js:80694:38)
    at checkExpressionCached (/workspaces/TypeScript/built/local/run.js:80358:38)
    at checkDeclarationInitializer (/workspaces/TypeScript/built/local/run.js:80375:23)
    at getTypeForVariableLikeDeclaration (/workspaces/TypeScript/built/local/run.js:56604:74)
    at getTypeForBindingElementParent (/workspaces/TypeScript/built/local/run.js:56312:61)
    at getNarrowedTypeOfSymbol (/workspaces/TypeScript/built/local/run.js:72373:46)
    at checkIdentifier (/workspaces/TypeScript/built/local/run.js:72497:24)
    at checkExpressionWorker (/workspaces/TypeScript/built/local/run.js:80745:28)
    at checkExpression (/workspaces/TypeScript/built/local/run.js:80694:38)
    at checkExpressionWithContextualType (/workspaces/TypeScript/built/local/run.js:80324:28)
    at checkDeclarationInitializer (/workspaces/TypeScript/built/local/run.js:80374:21)
*   at getTypeFromBindingElement (/workspaces/TypeScript/built/local/run.js:56990:81)
    at /workspaces/TypeScript/built/local/run.js:57040:109
    at [Object.map](http://object.map/) (/workspaces/TypeScript/built/local/run.js:455:29)
    at getTypeFromArrayBindingPattern (/workspaces/TypeScript/built/local/run.js:57040:35)
    at getTypeFromBindingPattern (/workspaces/TypeScript/built/local/run.js:57063:19)
    at getContextualTypeForInitializerExpression (/workspaces/TypeScript/built/local/run.js:73283:28)
    at getContextualType (/workspaces/TypeScript/built/local/run.js:73808:28)
    at getApparentTypeOfContextualType (/workspaces/TypeScript/built/local/run.js:73727:17)
    at checkArrayLiteral (/workspaces/TypeScript/built/local/run.js:74170:34)
    at checkExpressionWorker (/workspaces/TypeScript/built/local/run.js:80775:28)
    at checkExpression (/workspaces/TypeScript/built/local/run.js:80694:38)
    at checkExpressionCached (/workspaces/TypeScript/built/local/run.js:80358:38)
    at checkDeclarationInitializer (/workspaces/TypeScript/built/local/run.js:80375:23)
    at getTypeForVariableLikeDeclaration (/workspaces/TypeScript/built/local/run.js:56604:74)
    at getTypeForBindingElementParent (/workspaces/TypeScript/built/local/run.js:56312:61)
    at checkVariableLikeDeclaration (/workspaces/TypeScript/built/local/run.js:83464:34)
    at checkBindingElement (/workspaces/TypeScript/built/local/run.js:83605:20)
    at checkSourceElementWorker (/workspaces/TypeScript/built/local/run.js:86814:28)
    at checkSourceElement (/workspaces/TypeScript/built/local/run.js:86643:17)
    at Object.forEach (/workspaces/TypeScript/built/local/run.js:186:30)
    at checkVariableLikeDeclaration (/workspaces/TypeScript/built/local/run.js:83483:20)
    at checkVariableDeclaration (/workspaces/TypeScript/built/local/run.js:83600:13)
    at checkSourceElementWorker (/workspaces/TypeScript/built/local/run.js:86812:28)
    at checkSourceElement (/workspaces/TypeScript/built/local/run.js:86643:17)
    at Object.forEach (/workspaces/TypeScript/built/local/run.js:186:30)
    at checkVariableStatement (/workspaces/TypeScript/built/local/run.js:83611:16)
    at checkSourceElementWorker (/workspaces/TypeScript/built/local/run.js:86781:28)
    at checkSourceElement (/workspaces/TypeScript/built/local/run.js:86643:17)
    at Object.forEach (/workspaces/TypeScript/built/local/run.js:186:30)
    at checkSourceFileWorker (/workspaces/TypeScript/built/local/run.js:87018:20)
    at checkSourceFile (/workspaces/TypeScript/built/local/run.js:86983:13)
    at checkSourceFileWithEagerDiagnostics (/workspaces/TypeScript/built/local/run.js:87093:13)
    at getDiagnosticsWorker (/workspaces/TypeScript/built/local/run.js:87104:17)
    at Object.getDiagnostics (/workspaces/TypeScript/built/local/run.js:87070:24)
    at /workspaces/TypeScript/built/local/run.js:119101:85
    at runWithCancellationToken (/workspaces/TypeScript/built/local/run.js:119066:24)
    at getBindAndCheckDiagnosticsForFileNoCache (/workspaces/TypeScript/built/local/run.js:119084:20)
    at getAndCacheDiagnostics (/workspaces/TypeScript/built/local/run.js:119380:26)
    at getBindAndCheckDiagnosticsForFile (/workspaces/TypeScript/built/local/run.js:119081:20)
    at getSemanticDiagnosticsForFile (/workspaces/TypeScript/built/local/run.js:119078:61)
    at /workspaces/TypeScript/built/local/run.js:119014:24
    at Object.flatMap (/workspaces/TypeScript/built/local/run.js:519:25)
    at getDiagnosticsHelper (/workspaces/TypeScript/built/local/run.js:119010:56)
    at Object.getSemanticDiagnostics (/workspaces/TypeScript/built/local/run.js:119021:20)
    at Object.getPreEmitDiagnostics (/workspaces/TypeScript/built/local/run.js:117432:56)
    at Object.compileFiles (/workspaces/TypeScript/built/local/run.js:187832:42)
    at Object.compileFiles (/workspaces/TypeScript/built/local/run.js:190656:35)
    at new CompilerTest (/workspaces/TypeScript/built/local/run.js:200025:44)
    at Context.<anonymous> (/workspaces/TypeScript/built/local/run.js:199923:32)
    at callFn (/workspaces/TypeScript/node_modules/mocha/lib/runnable.js:366:21)
    at [Hook.run](http://hook.run/) (/workspaces/TypeScript/node_modules/mocha/lib/runnable.js:354:5)
    at next (/workspaces/TypeScript/node_modules/mocha/lib/runner.js:498:10)
    at Immediate._onImmediate (/workspaces/TypeScript/node_modules/mocha/lib/runner.js:559:5)
    at processImmediate (internal/timers.js:464:21)
    at process.callbackTrampoline (internal/async_hooks.js:130:17)'

**Comment by babakks** [comment]
@andrewbranch Could defining a new `CheckMode` a good idea? 🤔

**Comment by andrewbranch** [comment]
@babakks @weswigham speculated that probably only `CheckMode.Normal` should issue errors, so maybe you don’t even need a new one.

**Comment by Andarist** [comment]
I gotta say - I love your extensive descriptions here, @babakks ❤️

**Comment by babakks** [comment]
@andrewbranch I've submitted another PR with the `CheckMode` utilization approach. Could you please review it anytime?

**Comment by andrewbranch** [comment]
Looks like this was abandoned in favor of https://github.com/microsoft/TypeScript/pull/56753

**Pull request closed without merging**